File: Xit/XTFileChangesDataSource.m
#import "XTFileChangesDataSource.h"
#import "XTFileViewController.h"
#import "XTRepository+Parsing.h"
#import "Xit-Swift.h"

@interface XTFileChangesDataSource ()

@property NSMutableArray<XTFileChange*> *changes;

@end

@implementation XTFileChangesDataSource

- (void)reload
{
  [self.repository executeOffMainThread:^{
    NSMutableArray<XTFileChange*> *newChanges = self.winController.selectedModel.changes.mutableCopy;
    NSArray<NSSortDescriptor*> *pathDescriptors = @[
        [NSSortDescriptor sortDescriptorWithKey:@"path" ascending:YES] ];
    
    [newChanges sortUsingDescriptors:pathDescriptors];
    
    NSArray<NSString*>
        *newPaths = [newChanges valueForKey:@"path"],
        *oldPaths = [self.changes valueForKey:@"path"];
    NSOrderedSet<NSString*>
        *newSet = [NSOrderedSet orderedSetWithArray:newPaths],
        *oldSet = [NSOrderedSet orderedSetWithArray:oldPaths];
    
    NSIndexSet *deleteIndexes = [oldSet indexesOfObjectsPassingTest:
        ^BOOL(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
      return ![newSet containsObject:obj];
    }];
    NSIndexSet *addIndexes = [newSet indexesOfObjectsPassingTest:
        ^BOOL(NSString * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
      return ![oldSet containsObject:obj];
    }];
    
    NSIndexSet *changeIndexes = nil;
    NSMutableIndexSet *newChangeIndexes = nil;
    
    if (self.changes.count > 0) {
      newChangeIndexes = [NSMutableIndexSet indexSet];
      changeIndexes = [self.changes indexesOfObjectsPassingTest:
          ^BOOL(XTFileChange * _Nonnull obj,
                NSUInteger idx,
                BOOL * _Nonnull stop) {
        const NSUInteger newIndex = [newChanges indexOfObjectPassingTest:
            ^BOOL(XTFileChange * _Nonnull obj2,
                  NSUInteger idx,
                  BOOL * _Nonnull stop) {
          return [obj2.path isEqualTo:obj.path];
        }];
        
        if (newIndex == NSNotFound)
          return NO;
        
        XTFileChange *newChange = newChanges[newIndex];
        
        if ((newChange.change == obj.change) &&
            (newChange.unstagedChange == obj.unstagedChange))
          return NO;
        
        obj.change = newChange.change;
        obj.unstagedChange = newChange.unstagedChange;
        [newChangeIndexes addIndex:newIndex];
        return YES;
      }];
      [self.changes removeObjectsAtIndexes:deleteIndexes];
      [self.changes addObjectsFromArray:[newChanges objectsAtIndexes:addIndexes]];
      [self.changes sortUsingDescriptors:pathDescriptors];
    }
    else
      self.changes = newChanges;
    
    dispatch_async(dispatch_get_main_queue(), ^{
      if (self.outlineView.dataSource != self)
        return;
      [self.outlineView beginUpdates];
      if (deleteIndexes.count > 0)
        [self.outlineView removeItemsAtIndexes:deleteIndexes
                                      inParent:nil
                                 withAnimation:NSTableViewAnimationEffectFade];
      if (addIndexes.count > 0)
        [self.outlineView insertItemsAtIndexes:addIndexes
                                      inParent:nil
                                 withAnimation:NSTableViewAnimationEffectFade];
      [self.outlineView endUpdates];
      
      if (changeIndexes.count > 0) {
        NSIndexSet *allColumnIndexes = [NSIndexSet indexSetWithIndexesInRange:
            NSMakeRange(0, self.outlineView.numberOfColumns)];
        
        [self.outlineView reloadDataForRowIndexes:newChangeIndexes
                                    columnIndexes:allColumnIndexes];
      }
    });
  }];
}

- (BOOL)isHierarchical
{
  return NO;
}

- (XTFileChange*)fileChangeAtRow:(NSInteger)row
{
  if (row >= self.changes.count)
    return nil;
  return self.changes[row];
}

- (NSString*)pathForItem:(id)item
{
  XTFileChange *change = (XTFileChange*)item;

  return change.path;
}

- (XitChange)changeForItem:(id)item
{
  return [[self class] transformDisplayChange:((XTFileChange*)item).change];
}

- (XitChange)unstagedChangeForItem:(id)item
{
  return [[self class]
      transformDisplayChange:((XTFileChange*)item).unstagedChange];
}

#pragma mark NSOutlineViewDataSource

- (NSInteger)outlineView:(NSOutlineView*)outlineView
    numberOfChildrenOfItem:(id)item
{
  return self.changes.count;
}

- (id)outlineView:(NSOutlineView*)outlineView
            child:(NSInteger)index
           ofItem:(id)item
{
  return (index < self.changes.count) ? self.changes[index] : nil;
}

- (BOOL)outlineView:(NSOutlineView*)outlineView isItemExpandable:(id)item
{
  return NO;
}

- (id)outlineView:(NSOutlineView*)outlineView
    objectValueForTableColumn:(NSTableColumn*)tableColumn
                       byItem:(id)item
{
  return ((XTFileChange*)item).path;
}

@end

File: Xit/XTFileChangesDataSource.swift
import Foundation

class XTFileChangesDataSource: FileListDataSourceBase
{
  var changes = [FileChange]()
  var wasInStaging: Bool = false
  
  func doReload(_ model: FileChangesModel?)
  {
    var newChanges = model?.changes ?? [FileChange]()
    
    newChanges.sort { $0.path < $1.path }
    
    let newPaths = newChanges.map { $0.path }
    let oldPaths = changes.map { $0.path }
    let newSet = NSOrderedSet(array: newPaths)
    let oldSet = NSOrderedSet(array: oldPaths)
    
    let deleteIndexes = oldSet.indexes(options: []) {
      (path, _, _) in !newSet.contains(path) }
    let addIndexes = newSet.indexes(options: []) {
      (path, _, _) in !oldSet.contains(path) }
    var newChangeIndexes = IndexSet()
    
    objc_sync_enter(self)
    defer { objc_sync_exit(self) }
    
    if changes.isEmpty {
      changes = newChanges
    }
    else {
      changes.forEach {
        (change) in
        guard let newIndex = newChanges.index(where: {
          (newChange) in
          newChange.path == change.path &&
          ((newChange.change != change.change) ||
           (newChange.unstagedChange != change.unstagedChange))
        })
        else { return }
        
        let newChange = newChanges[newIndex]
        
        change.change = newChange.change
        change.unstagedChange = newChange.unstagedChange
        newChangeIndexes.insert(newIndex)
      }
      changes.removeObjects(at: deleteIndexes)
      changes.append(contentsOf: newChanges.objects(at: addIndexes))
      changes.sort { $0.path < $1.path }
    }
    
    DispatchQueue.main.async {
      [weak self] in
      guard let outlineView = self?.outlineView,
            self === outlineView.dataSource
      else { return }
      let selectedRow = outlineView.selectedRow
      let selectedChange = self?.fileChange(at: selectedRow)
      
      outlineView.beginUpdates()
      if !deleteIndexes.isEmpty {
        outlineView.removeItems(at: deleteIndexes,
                                inParent: nil,
                                withAnimation: .effectFade)
      }
      if !addIndexes.isEmpty {
        outlineView.insertItems(at: addIndexes,
                                inParent: nil,
                                withAnimation: .effectFade)
      }
      outlineView.endUpdates()
      
      if newChangeIndexes.count > 0 {
        // Have to construct an NSIndexSet and then convert to IndexSet
        // because the compiler doesn't recognize the constructor.
        let range = NSRange(location: 0, length: outlineView.numberOfColumns)
        let allColumnIndexes = NSIndexSet(indexesIn: range) as IndexSet
        
        outlineView.reloadData(forRowIndexes: newChangeIndexes,
                               columnIndexes: allColumnIndexes)
      }
      self?.reselect(change: selectedChange, oldRow: selectedRow)
      self?.updateStagingIcons()
    }
  }
  
  func updateStagingIcons()
  {
    if wasInStaging != controller.isStaging {
      let stagedIndex = outlineView.column(withIdentifier:
        FileViewController.ColumnID.staged)
      
      for (index, change) in changes.enumerated() {
        if let stagedCell = outlineView.view(atColumn: stagedIndex, row: index,
                                             makeIfNecessary: false)
                            as? TableButtonView {
          controller.updateTableButton(stagedCell.button,
                                       change: change.change,
                                       otherChange: change.unstagedChange)
        }
      }
      
      wasInStaging = controller.isStaging
    }
  }
  
  func reselect(change: FileChange?, oldRow: Int)
  {
    guard let oldChange = change
    else {
      if outlineView.selectedRowIndexes.isEmpty {
        outlineView.selectRowIndexes(IndexSet(integer: 0),
                                     byExtendingSelection: false)
      }
      return
    }
    var newRow = 0
    
    if let oldRowChange = fileChange(at: oldRow),
       oldRowChange.path == oldChange.path {
      newRow = oldRow
    }
    else {
      if let matchRow = changes.index(where: { $0.path == oldChange.path }) {
        newRow = matchRow
      }
    }
    outlineView.selectRowIndexes(NSIndexSet(index: newRow) as IndexSet,
                                 byExtendingSelection: false)
  }
}

extension XTFileChangesDataSource: FileListDataSource
{
  var hierarchical: Bool { return false }

  func reload()
  {
    let model = repoController.selectedModel
    
    taskQueue?.executeOffMainThread {
      [weak self] in
      self?.doReload(model)
    }
  }
  
  func fileChange(at row: Int) -> FileChange?
  {
    objc_sync_enter(self)
    defer { objc_sync_exit(self) }

    guard (row >= 0) && (row < changes.count)
    else { return nil }
    
    return row < changes.count ? changes[row] : nil
  }
  
  func path(for item: Any) -> String
  {
    return (item as? FileChange)?.path ?? ""
  }
  
  func change(for item: Any) -> DeltaStatus
  {
    guard let fileChange = item as? FileChange
    else { return .unmodified }
    
    return type(of: self).transformDisplayChange(fileChange.change)
  }
  
  func unstagedChange(for item: Any) -> DeltaStatus
  {
    guard let fileChange = item as? FileChange
    else { return .unmodified }
    
    return type(of: self).transformDisplayChange(fileChange.unstagedChange)
  }
}

extension XTFileChangesDataSource: NSOutlineViewDataSource
{
  func outlineView(_ outlineView: NSOutlineView,
                   numberOfChildrenOfItem item: Any?) -> Int
  {
    objc_sync_enter(self)
    defer { objc_sync_exit(self) }
    return changes.count
  }

  func outlineView(_ outlineView: NSOutlineView,
                   child index: Int,
                   ofItem item: Any?) -> Any
  {
    return (index < changes.count) ? changes[index] : FileChange(path: "")
  }

  func outlineView(_ outlineView: NSOutlineView,
                   isItemExpandable item: Any) -> Bool
  {
    return false
  }

  func outlineView(_ outlineView: NSOutlineView,
                   objectValueFor tableColumn: NSTableColumn?,
                   byItem item: Any?) -> Any?
  {
    return (item as? FileChange)?.path
  }
}
